File: src/main/java/com/github/sorusclient/client/InterfaceManager.java
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client;

import java.util.HashMap;
import java.util.Map;

public class InterfaceManager {

    private static final Map<Class<?>, Object> interfaces = new HashMap<>();

    public static void register(Object implementor) {
        interfaces.put(implementor.getClass().getInterfaces()[0], implementor);
    }

    public static <T> T get(Class<T> clazz) {
        return (T) interfaces.get(clazz);
    }

}

File: src/main/kotlin/com/github/sorusclient/client/InterfaceManager.kt
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client

object InterfaceManager {

    val interfaces: MutableMap<Class<*>, Any> = HashMap()

    fun <T> register(implementor: T) {
        interfaces[implementor!!::class.java.interfaces[0]] = implementor
    }

    inline fun <reified T> get(): T {
        return interfaces[T::class.java] as T
    }


}